diff --git a/src/interpreter/llvm/src/lib/CodeGen/LiveDebugValues/VarLocBasedImpl.cpp b/src/interpreter/llvm/src/lib/CodeGen/LiveDebugValues/VarLocBasedImpl.cpp
index 1e6d65c189..40b5e2a762 100644
--- a/src/interpreter/llvm/src/lib/CodeGen/LiveDebugValues/VarLocBasedImpl.cpp
+++ b/src/interpreter/llvm/src/lib/CodeGen/LiveDebugValues/VarLocBasedImpl.cpp
@@ -626,7 +626,7 @@ private:
     unsigned getRegIdx(Register Reg) const {
       for (unsigned Idx = 0; Idx < Locs.size(); ++Idx)
         if (Locs[Idx].Kind == MachineLocKind::RegisterKind &&
-            Locs[Idx].Value.RegNo == Reg)
+            (unsigned)Locs[Idx].Value.RegNo == Reg)
           return Idx;
       llvm_unreachable("Could not find given Reg in Locs");
     }
