diff --git a/src/core/unix/src/TUnixSystem.cxx b/src/core/unix/src/TUnixSystem.cxx
index 5b9574ef22..29d79f37ad 100644
--- a/src/core/unix/src/TUnixSystem.cxx
+++ b/src/core/unix/src/TUnixSystem.cxx
@@ -4613,7 +4613,7 @@ static const char *DynamicPath(const char *newpath = 0, Bool_t reset = kFALSE)
       dynpath += "/lib/x86_64-linux-gnu:/usr/local/lib64:/usr/lib64:/lib64:";
 #else
       // trick to get the system search path
-      std::string cmd("LD_DEBUG=libs LD_PRELOAD=DOESNOTEXIST ls 2>&1");
+      std::string cmd("LD_LIBRARY_PATH=${LD_LIBRARY_PATH:-DOESNOTEXIST} LD_DEBUG=libs LD_PRELOAD=DOESNOTEXIST ls 2>&1");
       FILE *pf = popen(cmd.c_str (), "r");
       std::string result = "";
       char buffer[128];
