diff --git a/src/core/textinput/src/textinput/StreamReaderWin.h b/src/core/textinput/src/textinput/StreamReaderWin.h
index 074a52d799..99d204af3f 100644
--- a/src/core/textinput/src/textinput/StreamReaderWin.h
+++ b/src/core/textinput/src/textinput/StreamReaderWin.h
@@ -16,7 +16,10 @@
 #define TEXTINPUT_STREAMREADERWIN_H
 
 #include "textinput/StreamReader.h"
-#include <Windows.h>
+
+#include <stdio.h>
+#include <intrin.h>
+#include <windef.h>
 
 namespace textinput {
   // Windows console and pipe input
diff --git a/src/core/textinput/src/textinput/TerminalDisplayWin.h b/src/core/textinput/src/textinput/TerminalDisplayWin.h
index 8aaceb23df..2447cd4e69 100644
--- a/src/core/textinput/src/textinput/TerminalDisplayWin.h
+++ b/src/core/textinput/src/textinput/TerminalDisplayWin.h
@@ -17,7 +17,10 @@
 #define TEXTINPUT_TERMINALDISPLAYWIN_H
 
 #include "textinput/TerminalDisplay.h"
-#include <Windows.h>
+#include <stdio.h>
+#include <intrin.h>
+#include <windef.h>
+
 
 namespace textinput {
   // Output to a Windows console or pipe.
