diff --git a/src/interpreter/cling/lib/Interpreter/DeclUnloader.cpp b/src/interpreter/cling/lib/Interpreter/DeclUnloader.cpp
index 25c01b1ef3..34b842583f 100644
--- a/src/interpreter/cling/lib/Interpreter/DeclUnloader.cpp
+++ b/src/interpreter/cling/lib/Interpreter/DeclUnloader.cpp
@@ -335,6 +335,7 @@ namespace {
     StaticVarCollector(FunctionDecl* FD, Vars& V) : m_V(V) {
       TraverseStmt(FD->getBody());
     }
+    bool shouldVisitImplicitCode() const { return true; }
     bool VisitDeclStmt(DeclStmt* DS) {
       for (DeclStmt::decl_iterator I = DS->decl_begin(), E = DS->decl_end();
            I != E; ++I)
