diff --git a/src/interpreter/cling/lib/Interpreter/DeclUnloader.cpp b/src/interpreter/cling/lib/Interpreter/DeclUnloader.cpp
index 467139e2d8..69d94e37b8 100644
--- a/src/interpreter/cling/lib/Interpreter/DeclUnloader.cpp
+++ b/src/interpreter/cling/lib/Interpreter/DeclUnloader.cpp
@@ -513,6 +513,7 @@ bool DeclUnloader::VisitRedeclarable(clang::Redeclarable<T>* R, DeclContext* DC)
       StaticVarCollector(FunctionDecl* FD, Vars& V) : m_V(V) {
         TraverseStmt(FD->getBody());
       }
+      bool shouldVisitImplicitCode() const { return true; }
       bool VisitDeclStmt(DeclStmt* DS) {
         for(DeclStmt::decl_iterator I = DS->decl_begin(), E = DS->decl_end();
             I != E; ++I)
